titus-server-master/src/main/java/com/netflix/titus/master/scheduler/constraint/OpportunisticCpuConstraint.java

```diff
@@ -0,0 +1,106 @@
+
+@Experimental(deadline = "10/2019")
+@Singleton
+public class OpportunisticCpuConstraint implements SystemConstraint {
+    public static final String NAME = OpportunisticCpuConstraint.class.getSimpleName();
+
+    private static final Result VALID = new Result(true, null);
+    private static final Result NO_OPPORTUNISTIC_CPUS = new Result(false, "The machine does not have opportunistic CPUs available");
+    private static final Result AVAILABILITY_NOT_LONG_ENOUGH = new Result(false, "CPU availability on the machine will not last for long enough");
+    private static final Result NO_RUNTIME_PREDICTION = new Result(false, "Task requested opportunistic CPUs without a runtime prediction");
+    private static final Result NOT_ENOUGH_OPPORTUNISTIC_CPUS = new Result(false, "The machine does not have enough opportunistic CPUs available");
+
+    private final SchedulerConfiguration configuration;
+    private final TaskCache taskCache;
+    private final OpportunisticCpuCache opportunisticCpuCache;
+    private final TitusRuntime titusRuntime;
+
+    @Inject
+    public OpportunisticCpuConstraint(SchedulerConfiguration configuration, TaskCache taskCache, OpportunisticCpuCache opportunisticCpuCache, TitusRuntime titusRuntime) {
+        this.configuration = configuration;
+        this.taskCache = taskCache;
+        this.opportunisticCpuCache = opportunisticCpuCache;
+        this.titusRuntime = titusRuntime;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public Result evaluate(TaskRequest taskRequest, VirtualMachineCurrentState targetVM, TaskTrackerState taskTrackerState) {
+        Preconditions.checkArgument(taskRequest instanceof V3QueueableTask, "opportunistic CPUs can only be used with V3QueueableTask");
+        V3QueueableTask request = (V3QueueableTask) taskRequest;
+
+        if (!request.isCpuOpportunistic()) {
+            return VALID;
+        }
+
+        Optional<Duration> runtimePrediction = request.getRuntimePrediction();
+        if (!runtimePrediction.isPresent()) {
+            titusRuntime.getCodeInvariants().inconsistent("Task requested opportunistic CPUs without a runtime prediction");
+            return NO_RUNTIME_PREDICTION;
+        }
+
+        String agentId = SchedulerUtils.getAttributeValueOrEmptyString(targetVM, configuration.getInstanceAttributeName());
+        if (StringExt.isEmpty(agentId)) {
+            titusRuntime.getCodeInvariants().inconsistent("No machine id attribute filled by Fenzo");
+            return NO_OPPORTUNISTIC_CPUS;
+        }
+
+        Optional<OpportunisticCpuAvailability> availabilityOpt = opportunisticCpuCache.findAvailableOpportunisticCpus(agentId);
+        if (!availabilityOpt.isPresent()) {
+            return NO_OPPORTUNISTIC_CPUS;
+        }
+
+        Duration taskRuntime = runtimePrediction.get();
+        Instant now = Instant.ofEpochMilli(titusRuntime.getClock().wallTime());
```

Comment:
agree, but opted for readability here, since it's a code path that's only triggered for a few tasks that have duration predictions, and a few agents that have opportunistic CPUs